Clients/Xamarin.Interactive.Client/CodeAnalysis/EvaluationService.cs

```diff
@@ -0,0 +1,510 @@
+            var model = new EvaluationModel ();
+            var cells = await GetAllCodeCellsAsync ();
+
+            var targetCellIndex = evaluateAll
+                ? -1
+                : cells.FindIndex (cell => cell.CodeCellId == targetCodeCellId);
+
+            // we're either evaluating all cells head to tail or we failed to
+            // find the target cell in the list; either way we're bailing early
+            if (targetCellIndex < 0) {
+                if (!evaluateAll && targetCodeCellId != default)
+                    throw new KeyNotFoundException (
+                        $"{nameof (targetCodeCellId)} '{targetCodeCellId}' not found");
+
+                model.ShouldResetAgentState = true;
+                model.CellsToEvaluate.AddRange (cells);
+                return model;
+            }
+
+            // otherwise, starting with our target cell and working backwards to the,
+            // head of the cell list figure out cells that need to be evaluated.
+            for (var i = targetCellIndex; i >= 0; i--) {
+                var cell = cells [i];
+
+                var isTargetCell = targetCellIndex == i;
+                var shouldEvaluate = isTargetCell || cell.IsEvaluationCandidate;
+
+                if (!shouldEvaluate && workspace.ShouldInvalidateCellBuffer (cell.CodeCellId)) {
+                    // a trick to force Roslyn into invalidating the tree it's holding on
+                    // to representing code pulled in via any #load directives in the cell.
+                    cell.Buffer.Invalidate ();
+                    shouldEvaluate = true;
+                }
+
+                if (shouldEvaluate) {
+                    model.ShouldResetAgentState |= i == 0;
+                    model.CellsToEvaluate.Insert (0, cell);
+                }
+            }
+
+            // now look at all cells after our target cell; if any of them have been
+            // evaluated before, we also want to re-evaluate those since they may
+            // depend on state in previous cells which will become invalidated.
+            for (var i = targetCellIndex + 1; i < cells.Count; i++) {
+                var cell = cells [i];
+
+                // if a cell was previously run but resulted in an agent termination,
+                // we do not want to automatically re-run that cell; the user must
+                // explicitly re-run terminated cells (which would be handled in the
+                // target->head walk above).
+                if (cell.AgentTerminatedWhileEvaluating)
+                    break;
+
+                // otherwise if we've evaluated this cell before, we should do so again
+                if (cell.EvaluationCount > 0)
+                    model.CellsToEvaluate.Add (cell);
+
+                // FIXME: this is where we did codeCellState.View.IsOutdated = true;
+                // but I do not know why we did that yet. Let this be a clue to future
+                // self. -abock, 2018-03-07
```

Comment:
Am I missing some nuance here? Every cell after the target cell should be marked outdated, right? In the UI this caused the results to look faded.

Reply:
Yeah, but I do not think we need to do that in `EvaluationService`. I wanted to omit it until we went to tackle it on the UI side. Remember that WPVM was an awful mixture of UI and state. `IsOutdated` would literally mess with the view.

Reply:
Absolutely. Was just trying to answer the "why". I do think that sort of state would be better managed fully on the client side.